edb/server/protocol/ext/auth.py

```diff
@@ -0,0 +1,282 @@
+
+from typing import Any, Type, TypeVar, TYPE_CHECKING
+from jwcrypto import jwt, jwk
+
+from edb import errors
+from edb.common import debug
+from edb.common import markup
+from edb.server.config import ops
+
+
+# TODO get this from edb.server.config.ops directly?
+SettingsMap = immutables.Map[str, ops.SettingValue]
+
+
+# Base class for OAuth 2 Providers
+class BaseProvider:
+    def __init__(self, name: str, client_id: str, client_secret: str):
+        self.name = name
+        self.client_id = client_id
+        self.client_secret = client_secret
+
+    def get_code_url(self) -> str:
+        raise NotImplementedError
+
+    async def exchange_code(self, code: str) -> str:
+        raise NotImplementedError
+
+    async def fetch_user_info(self, token: str) -> dict:
+        raise NotImplementedError
+
+
+# OAuth 2 Client HTTP Endpoints
+
+
+def redirect_to_auth_provider(
+    response,
+    redirect_uri: str,
+    iss: str,
+    provider_name: str,
+    db_config: SettingsMap,
+):
+    client_id, client_secret = _get_client_credentials(provider_name, db_config)
+    provider = _get_provider(
+        name=provider_name,
+        client_id=client_id,
+        client_secret=client_secret,
+    )
+
+    signing_key = _get_auth_signing_key(db_config)
+    state_token = _make_signed_token(
+        iss=iss, provider=provider_name, key=signing_key
+    )
+
+    auth_url = provider.get_code_url(
+        state=state_token.serialize(), redirect_uri=redirect_uri
+    )
+
+    response.status = http.HTTPStatus.FOUND
+    response.custom_headers["Location"] = auth_url
+    response.close_connection = True
+
+
+async def handle_auth_callback(request, response):
+    return
+
+
+async def handle_request(request, response, db, args):
+    try:
+        # Set up routing to the appropriate handler
+        if args[0] == "authorize":
+            provider_name: str | None = _get_search_param(
+                request.url.query.decode("ascii"), "provider"
+            )
+            if provider_name is None:
+                raise errors.InternalServerError("No provider specified")
+
+            redirect_uri = f"{_get_extension_path(request.url)}/callback"
+
+            return redirect_to_auth_provider(
+                response=response,
+                redirect_uri=redirect_uri,
+                iss=request.host.decode(),
+                provider_name=provider_name,
+                db_config=db.db_config,
+            )
+        elif args[0] == "callback":
+            await handle_auth_callback(request, response)
+        else:
+            raise errors.BackendError
+    except Exception as ex:
+        if debug.flags.server:
+            markup.dump(ex)
+
+        # XXX Fix this when LSP "location" objects are implemented
+        ex_type = errors.InternalServerError
+
+        _response_error(
+            response, http.HTTPStatus.INTERNAL_SERVER_ERROR, str(ex), ex_type
+        )
+
+
+def _response_error(response, status, message, ex_type):
```

Comment:
FWIW I suggest using keyword-only arguments here for clarity, e.g.

```suggestion
def _response_error(*, response, status, message, ex_type):
```

and then:

```python
        _response_error(
            response=response, 
            status=http.HTTPStatus.INTERNAL_SERVER_ERROR, 
            message=str(ex),
            ex_type=ex_type
        )
```

Also better to have verbs in method name; this method should be called `fail_with_error` or something.

Reply:
This is copy pasta from the other protocol handlers, but I can definitely model this here.